src/solve/moi.jl

```diff
@@ -0,0 +1,132 @@
+import MathOptInterface
+const MOI = MathOptInterface
+
+struct MOIOptimizationProblem{T,uType} <: MOI.AbstractNLPEvaluator
+    f::OptimizationFunction
+    u0::uType
+    J::Matrix{T}
+    H::Matrix{T}
+    cons_H::Vector{Matrix{T}}
+end
+
+MOI.eval_objective(moiproblem::MOIOptimizationProblem, x) = moiproblem.f.f(x)
+
+MOI.eval_constraint(moiproblem::MOIOptimizationProblem, g, x) = g .= moiproblem.f.cons(x)
+
+MOI.eval_objective_gradient(moiproblem::MOIOptimizationProblem, G, x) = moiproblem.f.grad(G, x)
+
+function MOI.eval_hessian_lagrangian(moiproblem::MOIOptimizationProblem{T}, h, x, σ, μ) where {T}
+    n = length(moiproblem.u0)
+    a = zeros(n, n)
+    moiproblem.f.hess(a, x)
+    if iszero(σ)
+        fill!(h, zero(T))
+    else
+        moiproblem.f.hess(moiproblem.H, x)
+        k = 0
+        for col in 1:n
+            for row in 1:col
+                k += 1
+                h[k] = σ * moiproblem.H[row, col]
+            end
+        end
+    end
+    if !isempty(μ) && !all(iszero, μ)
+        moiproblem.f.cons_h(moiproblem.cons_H, x)
+        for (μi, Hi) in zip(μ, moiproblem.cons_H)
+            k = 0
+            for col in 1:n
+                for row in 1:col
+                    k += 1
+                    h[k] += μi * Hi[row, col]
+                end
+            end
+        end
+    end
+    return
+end
+
+function MOI.eval_constraint_jacobian(moiproblem::MOIOptimizationProblem, j, x)
+    isempty(j) && return
+    moiproblem.f.cons_j === nothing && error("Use OptimizationFunction to pass the derivatives or automatically generate them with one of the autodiff backends")
+    n = length(moiproblem.u0)
+    moiproblem.f.cons_j(moiproblem.J, x)
+    for i in eachindex(j)
+        j[i] = moiproblem.J[i]
+    end
+end
+
+function MOI.jacobian_structure(moiproblem::MOIOptimizationProblem)
+    return Tuple{Int,Int}[(con, var) for con in 1:size(moiproblem.J,1) for var in 1:size(moiproblem.J,2)]
+end
+
+function MOI.hessian_lagrangian_structure(moiproblem::MOIOptimizationProblem)
+    return Tuple{Int,Int}[(row, col) for col in 1:length(moiproblem.u0) for row in 1:col]
+end
+
+function MOI.initialize(moiproblem::MOIOptimizationProblem, requested_features::Vector{Symbol})
+    for feat in requested_features
+        if !(feat in MOI.features_available(moiproblem))
+            error("Unsupported feature $feat")
+            # TODO: implement Jac-vec and Hess-vec products
```

Comment:
The Hess-vec is the `f.hv` function. What Jacobian? The constraint Jacobian?

Reply:
The Jacobian the derivative of constraints with respect to the variables of the problem. Yes, the constraint Jacobian.

Reply:
Yes, I limited this PR to just a rebase + fix of https://github.com/SciML/GalacticOptim.jl/pull/37 so this was copy-pasted from there.
The jacobian at row `i` and column `j` is the derivative of the `i`th constraint wrt the `j`th variable.